src/go/rpk/pkg/cli/topic/describe_storage.go

```diff
@@ -0,0 +1,258 @@
+			defer cl.Close()
+
+			adminCl, err := admin.NewClient(fs, p)
+			out.MaybeDie(err, "unable to initialize admin client: %v", err)
+
+			topic := args[0]
+			req := kmsg.NewPtrMetadataRequest()
+			reqTopic := kmsg.NewMetadataRequestTopic()
+			reqTopic.Topic = kmsg.StringPtr(topic)
+			req.Topics = append(req.Topics, reqTopic)
+
+			resp, err := req.RequestWith(context.Background(), cl)
+			out.MaybeDie(err, "unable to request topic metadata: %v", err)
+			if len(resp.Topics) != 1 {
+				out.Die("metadata response returned %d topics when we asked for 1", len(resp.Topics))
+			}
+			topicMeta := resp.Topics[0]
+			if err := kerr.ErrorForCode(topicMeta.ErrorCode); err != nil {
+				out.Die("error requesting topic metadata for %q: %v", topic, err)
+			}
+
+			var report []Status
+			for _, pa := range topicMeta.Partitions {
+				st, err := adminCl.CloudStorageStatus(context.Background(), topic, strconv.Itoa(int(pa.Partition)))
+				out.MaybeDie(err, "unable to request cloud storage status of topic %q, partition %v. make sure you have tiered storage to use this command: %v", topic, pa.Partition, err)
+				report = append(report, Status{pa.Partition, st})
+			}
+			if len(report) == 0 {
+				out.Die("error: could not get any status report from cloud storage.")
+			}
+			// By default, if neither are specified, we opt in to
+			// the size section only
+			if !summary && !size && !sync && !count && !offset {
+				summary, size = true, true
+			}
+			if all {
+				summary, size, sync, count, offset = true, true, true, true, true
+			}
+			var sections int
+			for _, b := range []bool{summary, size, sync, count, offset} {
+				if b {
+					sections++
+				}
+			}
+			withSections := sections > 1
+
+			header("SUMMARY", summary, withSections, func() {
+				tw := out.NewTabWriter()
+				defer tw.Flush()
+				tw.PrintColumn("NAME", topic)
+				if topicMeta.IsInternal {
+					tw.PrintColumn("INTERNAL", topicMeta.IsInternal)
+				}
+				tw.PrintColumn("PARTITIONS", len(topicMeta.Partitions))
+				if len(topicMeta.Partitions) > 0 {
+					p0 := &topicMeta.Partitions[0]
+					tw.PrintColumn("REPLICAS", len(p0.Replicas))
+				}
+				// We can safely pick the value of the first partition since
+				// the cloud storage mode doesn't change between partitions of
+				// the same topic.
+				tw.PrintColumn("CLOUD-STORAGE-MODE", report[0].CloudStatus.CloudStorageMode)
+
+				lastUpload := getLastUpload(report)
+				tw.PrintColumn("LAST-UPLOAD", humanReadable(lastUpload, human, "time"))
+			})
+
+			header("LOG-SIZE", size, withSections, func() {
+				tw := out.NewTable("PARTITION", "TOTAL", "CLOUD", "LOCAL")
+				defer tw.Flush()
+				for _, r := range report {
+					tw.Print(
+						r.Partition,
+						humanReadable(r.CloudStatus.TotalLogSize, human, "size"),
+						humanReadable(r.CloudStatus.CloudLogSize, human, "size"),
+						humanReadable(r.CloudStatus.LocalLogSize, human, "size"),
+					)
+				}
+			})
+
+			header("SYNC", sync, withSections, func() {
```

Comment:
I can't think of a better title for this, it's the most confusing section but this all makes sense when reading it in context.